src/java.base/share/classes/jdk/internal/util/ReferencedKeySet.java

```diff
@@ -0,0 +1,205 @@
+ * set.add(40_000_001L);
+ * set.add(40_000_002L);
+ * set.add(40_000_003L);
+ * set.add(40_000_004L);
+ * }
+ *
+ * @implNote Care must be given that the backing map does replacement by
+ * replacing the value in the map entry instead of deleting the old entry and
+ * adding a new entry, otherwise replaced entries may end up with a strongly
+ * referenced key. {@link HashMap} and {@link ConcurrentHashMap} are known
+ * to be safe.
+ *
+ * @param <T> the type of elements maintained by this set
+ */
+public final class ReferencedKeySet<T> extends AbstractSet<T> {
+    /**
+     * Backing {@link ReferencedKeySet} map.
+     */
+    final ReferencedKeyMap<T, ReferenceKey<T>> map;
+
+    /**
+     * Private constructor.
+     *
+     * @param map     backing map
+     */
+    private ReferencedKeySet(ReferencedKeyMap<T, ReferenceKey<T>> map) {
+        this.map = map;
+    }
+
+    /**
+     * Create a new {@link ReferencedKeySet} elements.
+     *
+     * @param isSoft          true if {@link SoftReference} elements are to
+     *                        be used, {@link WeakReference} otherwise.
+     * @param supplier        {@link Supplier} of the backing map
+     *
+     * @return a new set with {@link Reference} elements
+     *
+     * @param <E> the type of elements maintained by this set
+     */
+    public static <E> ReferencedKeySet<E>
+    create(boolean isSoft, Supplier<Map<ReferenceKey<E>, ReferenceKey<E>>> supplier) {
+        return create(isSoft, false, supplier);
+    }
+
+    /**
+     * Create a new {@link ReferencedKeySet} elements.
+     *
+     * @param isSoft          true if {@link SoftReference} elements are to
+     *                        be used, {@link WeakReference} otherwise.
+     * @param useNativeQueue  true if uses NativeReferenceQueue
+     *                        otherwise use {@link ReferenceQueue}.
+     * @param supplier        {@link Supplier} of the backing map
+     *
+     * @return a new set with {@link Reference} elements
+     *
+     * @param <E> the type of elements maintained by this set
+     */
+    public static <E> ReferencedKeySet<E>
+    create(boolean isSoft, boolean useNativeQueue, Supplier<Map<ReferenceKey<E>, ReferenceKey<E>>> supplier) {
+         return new ReferencedKeySet<>(ReferencedKeyMap.create(isSoft, useNativeQueue, supplier));
+    }
+
+    /**
+     * Removes enqueued weak references from set.
+     */
+    public void removeStaleReferences() {
+        map.removeStaleReferences();
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return map.keySet().iterator();
+    }
+
+    @Override
+    public int size() {
+        return map.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return map.isEmpty();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean contains(Object o) {
+        return map.containsKey((T)o);
+    }
+
+    @Override
+    public boolean add(T e) {
+        return intern(e) == null;
```

Comment:
This line is wrong, as `intern(…)` will never return `null`.

--------------------------------------------------------------------------------

This is the closest to the correct implementation,
```suggestion
        return !contains(e) & intern(e) == e;
```

but may incorrectly return `true` in case of the following data race, assuming `t1e == t2e`:

1. **Thread 1** calls `contains(t1e)` and gets `false`.
2. **Thread 2** calls `intern(t2e)` and successfully adds `t2e`.
3. **Thread 1** calls `intern(t1e)` and gets back `t2e`, which is `==` to `t1e`.
4. **Thread 1** returns `true`, even though it was **Thread 2** which modified the `ReferencedKeySet`.

Reply:
Good catch. Your solution might be correct but I think `!contains(e)` is redundant since that is how intern starts out.

```
   static <T> T intern(ReferencedKeyMap<T, ReferenceKey<T>> setMap, T key, UnaryOperator<T> interner) {
        T value = existingKey(setMap, key);
        if (value != null) {
            return value;
        }
        key = interner.apply(key);
        return internKey(setMap, key);
    }
```

Agree? So changing to `return intern(e) == e;` should be sufficient.

The other aspect of this is that `internKey` uses `putIfAbsent` which should prevent the race (assuming `ConcurrentHashMap`).

```
    /**
     * Attempt to add key to map.
     *
     * @param setMap    {@link ReferencedKeyMap} where interning takes place
     * @param key       key to add
     *
     * @param <T> type of key
     *
     * @return the old key instance if found otherwise the new key instance
     */
    private static <T> T internKey(ReferencedKeyMap<T, ReferenceKey<T>> setMap, T key) {
        ReferenceKey<T> entryKey = setMap.entryKey(key);
        T interned;
        do {
            setMap.removeStaleReferences();
            ReferenceKey<T> existing = setMap.map.putIfAbsent(entryKey, entryKey);
            if (existing == null) {
                return key;
            } else {
                // If {@code putIfAbsent} returns non-null then was actually a
                // {@code replace} and older key was used. In that case the new
                // key was not used and the reference marked stale.
                interned = existing.get();
                if (interned != null) {
                    entryKey.unused();
                }
            }
        } while (interned == null);
        return interned;
    }
```

Agree? Anyone else want to pipe in?

Reply:
ok, `intern(e) == e` is sufficient.

Reply:
Created https://bugs.openjdk.org/browse/JDK-8325255

Reply:
While `intern(e) == e` is (mostly) sufficient, it will return a false positive when `add(e)` is called and `e` is already present in the map, the correctest implementation would be some internal API in `ReferencedKeyMap` for implementing `ReferencedKeySet::add`.